### PR TITLE
Clear ambient inherited

### PR DESCRIPTION
**Pull request by kkuehlz**

### CODE REVIEW

**Comment by keur** [comment]
cc @anitazha

**Comment by yuwata** [comment]
`test-load-fragment` fails.

**Comment by yuwata** [comment]
What happens when the following is specified?
```
AmbientCapabilities=CAP_AAA
AmbientCapabilities=clear
AmbientCapabilities=CAP_BBB
AmbientCapabilities=
AmbientCapabilities=CAP_CCC
```

Also, I think DBus API needs to be updated.

**Comment by mrc0mmand** [comment]
```
==5697==ERROR: AddressSanitizer: stack-use-after-return on address 0x7fab7ca59d18 at pc 0x55ab9274bc57 bp 0x7ffd78d0c5f0 sp 0x7ffd78d0c5e8
READ of size 8 at 0x7fab7ca59d18 thread T0
    #0 0x55ab9274bc56 in config_parse_capability_set /build/build/../src/core/load-fragment.c:1386:25
    #1 0x55ab926e5dea in test_config_parse_capability_set /build/build/../src/test/test-load-fragment.c:582:13
    #2 0x55ab926e1040 in main /build/build/../src/test/test-load-fragment.c:791:9
    #3 0x7fab80ca7152 in __libc_start_main (/lib64/libc.so.6+0x27152)
    #4 0x55ab926e0e5d in _start (/build/build/test-load-fragment+0x2bae5d)

Address 0x7fab7ca59d18 is located in stack of thread T0 at offset 24 in frame
    #0 0x7fab815bf44f in utf8_encoded_valid_unichar /build/build/../src/basic/utf8.c:502

  This frame has 1 object(s):
    [32, 36) 'unichar' (line 503) <== Memory access at offset 24 underflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-use-after-return /build/build/../src/core/load-fragment.c:1386:25 in config_parse_capability_set
Shadow bytes around the buggy address:
  0x0ff5ef943350: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0ff5ef943360: f1 f1 f1 f1 00 f3 f3 f3 f1 f1 f1 f1 00 f3 f3 f3
  0x0ff5ef943370: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0ff5ef943380: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0ff5ef943390: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
=>0x0ff5ef9433a0: f5 f5 f5[f5]f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0ff5ef9433b0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0ff5ef9433c0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0ff5ef9433d0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0ff5ef9433e0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0ff5ef9433f0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==5697==ABORTING
```

**Comment by poettering** [comment]
hmm, is this really desirable this way? why not drop all ambient caps by default, and then allow AmbientCapabilities= as a method of passing them along?

**Comment by keur** [comment]
> hmm, is this really desirable this way? why not drop all ambient caps by default, and then allow AmbientCapabilities= as a method of passing them along?

I am fine doing it this way, but I'd prefer something more explicit, like `AmbientCapabilities=inherit`

**Comment by keur** [comment]
@poettering Yeah, I definitely agree. Having to specify dropping caps would be an opt in security feature, which is not good. Fixed and force pushed.

cc @yuwata

File: src/basic/capability-util.c [hunk 1]
```diff
@@ -138,6 +138,42 @@ int capability_ambient_set_apply(uint64_t set, bool also_inherit) {
         return 0;
 }
 
+int capability_ambient_set_clear(uint64_t set) {
```
Comment (poettering) [inline src/basic/capability-util.c]: hmm, iiuc the parameter specifies the ones to keep, right? hence call it "keep" or so, to make this clearer?

File: src/basic/capability-util.c [hunk 2]
```diff
@@ -138,6 +138,42 @@ int capability_ambient_set_apply(uint64_t set, bool also_inherit) {
...
+
+                if (prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_LOWER, i, 0, 0) < 0)
+                        return -errno;
+        }
+
+        if (cap_set_proc(tmp_cap) < 0)
+                return -errno;
+
+        return 0;
+}
```
Comment (poettering) [inline src/basic/capability-util.c]: shouldn't this be done in capability_ambient_set_apply()? i.e. that the function is changed to drop all ambient caps not specified and set all ambient caps specified? that would be much shorter and simpler and still do what you need, no?

**Comment by keur** [comment]
@poettering Changes made.

File: src/test/test-capability.c [hunk 3]
```diff
@@ -212,6 +212,14 @@ static void test_set_ambient_caps(void) {
         cap_free(caps);
 
         assert_se(prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_IS_SET, CAP_CHOWN, 0, 0) == 1);
+
+        assert_se(!capability_ambient_set_apply(0, true));
+        caps = cap_get_proc();
```
Comment (poettering) [inline src/test/test-capability.c:218]: please enclose this in assert_se() too

**Comment by poettering** [comment]
looks great! one more tiny fix

**Comment by keur** [comment]
@poettering Fixed and force pushed (also added the assert_se to the other call to cap_get_proc from the existing test)

File: src/test/test-capability.c [hunk 4]
```diff
@@ -207,11 +207,21 @@ static void test_set_ambient_caps(void) {
...
         assert(fv == CAP_SET);
         cap_free(caps);
 
         assert_se(prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_IS_SET, CAP_CHOWN, 0, 0) == 1);
+
+        assert_se(!capability_ambient_set_apply(0, true));
+        caps = cap_get_proc();
+        assert_se(caps);
+        assert_se(!cap_get_flag(caps, CAP_CHOWN, CAP_INHERITABLE, &fv));
+        assert(fv == CAP_CLEAR);
```
Comment (poettering) [inline src/test/test-capability.c]: all our tests should use assert_se() btw, i.e. an assert check that is not optimized away, after all our tests are supposed to be tests, and hence allowing them to be optimized away is besides the point...

Comment (kkuehlz) [reply src/test/test-capability.c]: makes sense. thanks for the explanation

**Comment by poettering** [comment]
thanks! lgtm, but let's merge that after the 244 release

**Pull request merged**